### ISSUE TITLE
Suffix package versions to work correctly with APT tools and mirrors

### ISSUE THREAD

**Issue by jgraichen** [saltstack/salt-auto-pack#90]
The salt packaging setup is rebuilding identical packages for different distributions with the same version and architecture tuple. This isn't supported by most apt and dpkg tooling. Debian repositories expect that a package with the same version and architecture to always be byte identical.

Salt builds an entirely new APT repository for each distribution, e.g.

* https://repo.saltproject.io/py3/ubuntu/20.04/amd64/latest
* https://repo.saltproject.io/py3/ubuntu/18.04/amd64/latest

Both repositories contain a `python3-contextvars` package in version `2.4-1` build for `amd64` but they are actually different packages (different checksums).

This isn't and immediate issue here as they are managed as separate repositories, but this isn't how most open-source and enterprise APT repositories work. Usually, there is a single repository serving different distributions, releases, and architectures.

For example, [Debian](http://ftp.debian.org/debian/dists/), serving all Debian distributions from the same repository, or PostgreSQL serving multiple Debian and Ubuntu released from the same repository: http://apt.postgresql.org/pub/repos/apt/dists/. Corporate mirrors often do the same, they mirror packages from multiple upstreams, merge them into a single repository, and serve packages for all distributions that are in use.

Unfortunately, because salt packages and dependencies are built again for each distribution and architecture but with the same version, salt packages _cannot_ be mirrored in any combined repository, for example, when trying to mirror the packages for Ubuntu 18.04 _and_ 20.04 into our mirror, it will complain:

> ERROR: unable to publish: unable to process packages: error linking file to /var/lib/aptly/data/public/cloud/pool/main/c/contextvars/python3-contextvars_2.4-1_all.deb: file already exists and is different

This makes it impossible to provide salt for multiple distributions, or a new version, in any single APT repository. This was already mentioned e.g. [here](https://github.com/saltstack/salt-pack/issues/594#issuecomment-435084441) but not changed.

---

A fix would be straightforward: Add a distribution suffix to each version of all packages that are built. For example, the PostgreSQL repository adds the following suffixes for the different distributions:

```
postgresql-14_14.2-1.pgdg18.04+1_amd64.deb
postgresql-14_14.2-1.pgdg18.04+1_arm64.deb
postgresql-14_14.2-1.pgdg20.04+1_amd64.deb
postgresql-14_14.2-1.pgdg20.04+1_arm64.deb
```

This way, versions are unique even between repositories. This is the common approach to handle this case, even by the Debian backports repository, which, for example, adds `~bpo11+1` for the backported `haproxy` package:

```
bullseye-backports: 2.4.15-1~bpo11+1
bookworm          : 2.4.15-1
```

It would be wonderful if salt could follow the same package building rules so that we could import these packages into our mirrors. This is especially important as the bundled packages in e.g. Debian and Ubuntu always have issues, from not being included at all (Ubuntu 20.04), to not being able to talk any current master due to changed in the wire protocol (Ubuntu 22.04). We already use e.g. `cloud-init` to automatically add our own APT repository, and to install salt and start the minion. Adding a single repository in `cloud-init` is effortless, it just needs a single line:

```
deb http://apt-host/ $RELEASE main
```

This will not work that easily with separate repositories for each distribution and architecture, as needed for directly using the salt repositories. We also use the mirror to control the salt updated, e.g. to ensure the master is updated before the minions (which use e.g. unattended-upgrades), by first upgrading the master, and then importing the new salt-minion package. This is currently not possible for multiple distributions or architectures.

---

I am not sure which is the correct repo here, but `salt-pack` and `salt-pack-py3` seems not to include the latest versions. Is this one the currently used one?

I'd be interested in trying to add the needed changes, at least to the build script.

**Comment by dmurphy18** [comment]
@jgraichen Yes, this issue is well-known as you have referenced.  The Salt Project is resolving this issue by moving to a self-contained Tiamat based package which includes all dependencies and Python 3 version used.

This allows the Salt Project to move forward with its code-base and still be able to support older platforms with new releases.  For example: Debian 9 is still supported, but uses Python 3.5 which is long end of life'd.  With Tiamat based packaging, a salt-minion can be provided which leverages Python 3.9.  That is a deb package can be provided using Tiamat internally rather than the current method of individual packages.

For example: with the soon to be released Ubuntu 22.04, it will be a deb package with Tiamat internally used.
Tiamat based packaging has been used for some years already on Arista and Juniper routers where only Python 2.7 with 'C' and POSIX environments only available.  Using Tiamat, and leveraging PEP 538 and 540, UTF-8 support with Python 3.7 was possible.

The Tiamat packaging has been in Beta for Linux distribution since currently published versions did not support pip installs (not an issue on routers), but with the upcoming Phosphorous release, that short coming has been addressed.

Part of the reasoning behind Tiamat packaging has been the dependency hell using OS dependencies packages, which typically took more time to resolve than the actual Salt OS package itself.  This usually manifested itself as the OS distribution aged and support for it fell off.

As for salt-pack and salt-pack-py3, they should be updated soon.

**Comment by dmurphy18** [comment]
@jgraichen Also note build everything on the OS platform to ensure that it is reproducible on that platform, hence checksum change. Sometimes a dependency has to be adjusted for a later OS platform, since its assumption might no longer be valid.  This has happened frequently, to my exasperation, hence built once and run everywhere has not been my experience.

**Comment by jgraichen** [comment]
@dmurphy18 Thanks a lot for that prompt update!

We did start testing the first Ubuntu 22.04 machines due to the upcoming release and were hit by the issue again. Are there already Debian/Ubuntu packages for the Tiamat binary that I begin to test in our environment?

We run with many custom states, grains, and modules too, how will they be affected? Will they run with the embedded version or will salt be a standalone binary, still using the system python to run modules?

> Also note build everything on the OS platform to ensure that it is reproducible on that platform, hence checksum change. Sometimes a dependency has to be adjusted for a later OS platform, since its assumption might no longer be valid.

That is no problem and not much different from what PostgreSQL is doing. All that would be needed is to append a suffix to the version, e.g. `2.4-1~18.04` when building on Ubuntu Bionic, and `2.4-1~20.04` when building on Ubuntu Focal.

> hence built once and run everywhere has not been my experience

I mainly see (and use) this with e.g. packages for Golang binaries, that often are very system-independent. Not something with Python, yes.

Debian and Ubuntu, they never rebuild a package if it has not been updated, or, when backporting a package, they also suffix them.

**Comment by dmurphy18** [comment]
@jgraichen Don't have 22.04 packages available for external use yet, but there are BETA versions for Ubuntu 20.04, just don't try pip install.  Tiamat uses internal version of Python 3.7 in the BETA versions, but should be internally Python 3.9 in Phosphorous release (Understand Py 3.10 in Ubuntu 22.04, but trying to standardize across platforms for this release). 

And yes, marking packages with a suffix is done '+ds-1' etc, with hindsight adding that version too would have been smart.  

Understand Debian and the desire to build or re-build a package such that it is bit for bit the same to ensure the same checksum, but unless you have a complete toolchain lockdown including OS version etc., not possible in this general case, but in past lives have done so with priority software back in the late 90's.

**Comment by jgraichen** [comment]
If the Tiamat package runs unmodified on any distribution, the suffix wouldn't be important anymore. I could just import the identical package into each dist.

Until then, would it be possible to suffix the normal packages you are still distributing with a dist suffix? This wouldn't fix your maintenance issues, but all the issues from people that have to import and mirror your packages.

Unfortunately, I have not found a way through the salt states here, or to run them without the actual salt master needed there. I'd be happy to assist or test any new packages through, or if you can point to where exactly packages are built, I can try to help too.

**Comment by dmurphy18** [comment]
They should come wrapped up as rpm's or deb's as appropriate for the platform.  The packages include things like default configuration files, systemD service files etc.

The nightly build versions are available from https://repo.saltproject.io/salt-dev/ but do NOT use in production environments as theses are not release versions and have not gone through through QA etc.

But can be used to gain familiarity.  There is no support and documentation is still being updated for release and not fully available yet, but examination of Tiamat, Tiamat-pip on GitLab have README.md associated with them.  Similarly the Native Minions documentation (no Tiamat-pip support) can also be used for an idea as to how salt-minions are used.

**Comment by jgraichen** [comment]
> Also note build everything on the OS platform to ensure that it is reproducible on that platform, hence checksum change. Sometimes a dependency has to be adjusted for a later OS platform, since its assumption might no longer be valid. This has happened frequently, to my exasperation, hence built once and run everywhere has not been my experience.

That is perfectly fine. Can you just append a suffix to each package? For example, when building `python3-zmq_17.1.2` for Ubuntu 18.04, append `~ubu18` to the version (`python3-zmq_17.1.2~ubu18`)?

This means versions are unique to the distribution, and we can import salt and its dependencies for Ubuntu 18.04, 20.04, 22.04, and Debian 10, and 11 into our mirror?
